ml/src/main/scala/frameless/ml/TypedTransformer.scala

```diff
@@ -0,0 +1,42 @@
+package frameless
+package ml
+
+import frameless.ops.SmartProject
+import org.apache.spark.ml.Transformer
+import shapeless.{Generic, HList}
+import shapeless.ops.hlist.{Prepend, Tupler}
+
+sealed trait TypedTransformer
+
+/**
+  * An AppendTransformer `transform` method takes as input a TypedDataset containing `Inputs` and
+  * return a TypedDataset with `Outputs` columns appended to the input TypedDataset.
+  */
+trait AppendTransformer[Inputs, Outputs, InnerTransformer <: Transformer] extends TypedTransformer {
+  val transformer: InnerTransformer
+
+  def transform[T, TVals <: HList, OutputsVals <: HList, OutVals <: HList, Out, F[_]](ds: TypedDataset[T])(
```

Comment:
@atamborrino to the best of my knowledge, when you execute transform, this is a lazy operation from Dataset to Dataset. I don't think we need to enclose it into the effectful `SparkDelay`. I think transform should just take a `TypedDataset[T]` and rerun a `TypedDataset[Out]`.

Reply:
I sampled a few mllib `Model`s and that looks right, although they may throw exceptions (`require` is used liberally).

Reply:
@imarios you're right, I had tested that but my test was apparently incorrect... I'll remove the SparkDelay for Transformers. 
@iravid hopefully the conditions of the `require`s are checked at compile-time when building the TypedTransformer.